src/tools/illink/test/Mono.Linker.Tests.Cases/CommandLine/Dependencies/CustomApplyPreserveStep.cs

```diff
@@ -0,0 +1,13 @@
+﻿namespace CustomStep
+{
+    public class CustomApplyPreserveStep : Mono.Linker.Steps.IStep
+    {
+        public void Process(Mono.Linker.LinkContext context)
+        {
+            var myType = context.GetType("Mono.Linker.Tests.Cases.CommandLine.CustomStepApplyPreserve/HasPreserveApplied");
+            context.Annotations.SetPreserve(myType, Mono.Linker.TypePreserve.Methods);
+            // Make sure we can set preserve multiple times
+            context.Annotations.SetPreserve(myType, Mono.Linker.TypePreserve.All);
```

Comment:
Calling `SetPreserve` once fails pre-change on the exception from https://github.com/dotnet/runtime/issues/103115. The double call to `SetPreserve` doesn't test the related issue, but I added it to help me understand the code and I thought it wouldn't hurt to leave it in since we don't have many custom steps tests.

This repro does those things in order:
- CustomApplyPreserveStep runs before MarkStep and sets the preserve on the type
- ProcessMarkedPending runs at the start of MarkStep and call SetAppliedPreserve
- The type is marked in Main() and ProcessType calls SetAppliedPreserve again